xclim/core/formatting.py

```diff
@@ -321,6 +323,84 @@ def update_history(
     return merged_history
 
 
+def update_xclim_history(func):
+    """Decorator that auto-generates and fills a xclim_history attribute.
+
+    The history is generated from the signature of the function and added to the first output.
+    """
+
+    @wraps(func)
+    def _call_and_add_history(*args, **kwargs):
+        """Call the function and then generate and add the xclim_history attr."""
+        outs = func(*args, **kwargs)
+
+        if isinstance(outs, tuple):
+            out = outs[0]
+        else:
+            out = outs
+
+        if not isinstance(out, (xr.DataArray, xr.Dataset)):
+            raise TypeError(
+                f"Decorated `update_xclim_history` received a non-xarray output from {func.__name__}."
+            )
+
+        da_list = [arg for arg in args if isinstance(arg, xr.DataArray)]
+        da_dict = {
+            name: arg for name, arg in kwargs.items() if isinstance(arg, xr.DataArray)
+        }
+
+        attr = update_history(
+            gen_call_string(func.__name__, *args, **kwargs),
+            *da_list,
+            new_name=out.name,
+            **da_dict,
+        )
+        out.attrs["xclim_history"] = attr
+        return outs
+
+    return _call_and_add_history
+
+
+def gen_call_string(funcname: str, *args, **kwargs):
```

Comment:
Ah nice ! This will be useful for provenance work as well.

Reply:
Open question: should we pass the func object instead so it can be inspected ? The advantage I see is that we could bind *args and print input parameters as **kwargs for clarity.

Reply:
haha. I began writing this taking in the function (or the signature), but realized there were no useful supplementary information?

One thing I tried to do was to split "positional-only" from "keyword" arguments according to the signature (instead of whether they are in *args or **kwargs). But I wasn't convinced that was interesting enough to justify needing a function.

Also, in the Indicator, we replace the _function_ name with the indicator's ID. It was easier to do this if the name was passed.

Reply:
What if we want to print the full path to the function, e.g. `xclim.atmost.growing_degree_days` ? Imagine we have two `average` indicators in two different modules, that could get confusing.

Reply:
Well, this would confirm my choice of taking a "funcname" as input! So the calling function could decide on what name to use. In the indicator, using the "registry_id" fulfills the need you are indicating since it is "unique".

Reply:
And I could update `update_xclim_history` to prepend  `func.__module__` to solve this in sdba too.